crates/matrix-sdk/src/room/edit.rs

```diff
@@ -0,0 +1,187 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Facilities to edit existing events.
+
+use ruma::{
+    events::{
+        room::message::{Relation, ReplacementMetadata, RoomMessageEventContentWithoutRelation},
+        AnyMessageLikeEvent, AnyMessageLikeEventContent, AnySyncMessageLikeEvent,
+        AnySyncTimelineEvent, AnyTimelineEvent, MessageLikeEvent, SyncMessageLikeEvent,
+    },
+    EventId,
+};
+use thiserror::Error;
+use tracing::{instrument, warn};
+
+use crate::Room;
+
+/// The new content that will replace the previous event's content.
+pub enum EditedContent {
+    /// The content is a `m.room.message`.
+    RoomMessage(RoomMessageEventContentWithoutRelation),
+}
+
+#[cfg(not(tarpaulin_include))]
+impl std::fmt::Debug for EditedContent {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::RoomMessage(_) => f.debug_tuple("RoomMessage").finish(),
+        }
+    }
+}
+
+impl EditedContent {
+    /// Return a human-readable type describing the content's type, for use in
+    /// error messages.
+    fn human_readable_type(&self) -> &'static str {
+        match self {
+            EditedContent::RoomMessage(_) => "room message",
+        }
+    }
+}
+
+/// An error occurring while editing an event.
+#[derive(Debug, Error)]
+pub enum EditError {
+    /// We tried to edit a state event, which is not allowed, per spec.
+    #[error("State events can't be edited")]
+    StateEvent,
+
+    /// We tried to edit an event which sender isn't the current user, which is
+    /// forbidden, per spec.
+    #[error("You're not the author of the event you'd like to edit.")]
+    NotAuthor,
+
+    /// We couldn't fetch the remote event with /room/event.
+    #[error("Couldn't fetch the remote event: {0}")]
+    Fetch(Box<crate::Error>),
+
+    /// We couldn't properly deserialize the target event.
+    #[error(transparent)]
+    Deserialize(#[from] serde_json::Error),
+
+    /// We tried to edit an event of type A with content of type B.
+    #[error("The original event type ({target}) isn't the same as the parameter's new content type ({new_content})")]
+    IncompatibleEditType {
+        /// The type of the target event.
+        target: &'static str,
+        /// The type of the new content.
+        new_content: &'static str,
+    },
+
+    /// While it's technically possible to edit this kind of events, this is not
+    /// yet supported by the SDK.
+    #[error("Edits for events of type {0} aren't supported yet")]
+    UnsupportedEventType(String),
```

Comment:
It seems like this error should already be catched by the `IncompatibleEditType` variant since all supported types have a variant in `EditedContent`.

Reply:
Good point! By rejiggering the order of these match statements, we could probably spare one error type indeed :+1: